scripts/gen_dockerfile.py

```diff
@@ -0,0 +1,250 @@
+#!/usr/bin/env python3
+
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generate a Dockerfile and helper files for a Python application."""
+
+import argparse
+import collections
+import functools
+import io
+import os
+import re
+import sys
+
+import yaml
+
+import validation_utils
+
+
+# Validate characters for dockerfile image names.
+#
+# This roots out obvious mistakes, the full gory details are here:
+# https://github.com/docker/distribution/blob/master/reference/regexp.go
+IMAGE_REGEX = re.compile(r"""(?x)
+    ^
+    [a-zA-Z0-9]          # First char must be alphanumeric
+    [a-zA-Z0-9-_./:@+]*  # Punctuation allowed after that
+    $
+""")
+
+# `entrypoint` is specified as free-form text parsed as a unix shell
+# command line, which limits the sanity checking possible.  We
+# disallow newlines and control characters which would break the
+# Dockerfile format.
+PRINTABLE_REGEX = re.compile(r"""^[^\x00-\x1f]*$""")
+
+# Map from app.yaml "python_version" to {python_version} in Dockerfile
+PYTHON_INTERPRETER_VERSION_MAP = {
+    '': '',  # == 2.7
+    '2': '',  # == 2.7
+    '3': '3.5',
+    '3.4': '3.4',
+    '3.5': '3.5',
+    '3.6': '3.6',
+}
+
+# File templates.
+# Designed to exactly match the current output of 'gcloud app gen-config'
+DOCKERFILE_TEMPLATE = """\
```

Comment:
I moved the string template into separate files, and still use regular string formatting.  Actually, I (re)copied the template data files from the 'data' directory of the gae-xrt-python package of the gcloud SDK, since we need to keep the two sets of things in sync for the near future.  I would link to this if we mirrored the SDK source to a public repository, but we don't.  I kept the file names and contents all the same, which required a little reorganizing of my code.